### DISCUSSION TITLE
Water97 fluid properties problem

### DISCUSSION THREAD

**Discussion by SomnusYu** [idaholab/moose#25321] [Q&A General]
Hi all,

I would like to use the partial derivative of sound to pressure in Water97FluidProperties, but I can not find the derivative function of sound. The only one that related to sound is **[Water97FluidProperties::c_from_p_T_template](https://mooseframework.inl.gov/docs/doxygen/modules/classWater97FluidProperties.html#aeb7baee814f46166c7888247dbf49764)(const T & [pressure](https://mooseframework.inl.gov/docs/doxygen/modules/namespaceNS.html#ae996468059a393b93fdc3afe09ef40b9), const T & [temperature](https://mooseframework.inl.gov/docs/doxygen/modules/namespaceNS.html#a72935b300cfc0975e3f43c66b3d98526)) const**. The derivative terms are not included in this function.

So, how to calculate the partial derivative of sound to pressure? Can anyone kindly tell me?

Best regards

**Comment by GiudGiud** [comment]
Hello

You'll need to add that definition, we have not added it.
Add this signature:
```
c_from_p_T(const Real pressure, const Real temperature, Real & c, Real & dc_dp, Real & dc_dT) const
```
to WaterFP

you may have to template it later on

Guillaume

**Reply by SomnusYu** [reply]
Thank you! But it seems so complicated to calculate the dc_dp! And I do not know the partial derivative of these terms **(just like d2gamma1_dpitau(pi, tau), d2gamma1_dtau2(pi, tau), d2gamma1_dpi2(pi, tau))** to pressure.

**Reply by cpgr** [reply]
Yeah, they are super tedious! These were mostly coded before AD was available in MOOSE. You could see if you could 'cheat' and use AD to compute the derivatives - like 

https://github.com/idaholab/moose/blob/516ad1215a7f38384beab4549202215bd396ccf2/modules/fluid_properties/src/fluidproperties/Water97FluidProperties.C#L689C25-L689C35

**Reply by SomnusYu** [reply]
So, if I use this way, I need to update my moose version, is that right?

**Reply by GiudGiud** [reply]
If it s really really old yes.
If not then AD has been around for years.

**Reply by SomnusYu** [reply]
OK, thank you very much!
